packages/headless/src/features/insight-user-actions/insight-user-actions-preprocessing.ts

```diff
@@ -0,0 +1,277 @@
+    });
+
+    if (ticketCreationActionIndex !== -1) {
+      session.actions.splice(
+        ticketCreationActionIndex,
+        0,
+        ticketCreationAction
+      );
+    }
+    timeline.session = session;
+  } else if (ticketCreationDate < session.start) {
+    timeline.followingSessions.push(session);
+  } else {
+    timeline.precedingSessions.push(session);
+  }
+};
+
+/**
+ * Checks whether the action should be excluded from the session based on the excluded custom actions and empty searches.
+ * @param action {UserAction} - The action to check
+ * @param excludedCustomActions {string[]} - The custom actions to exclude
+ * @returns {boolean}
+ */
+const shouldExcludeAction = (
+  action: UserAction,
+  excludedCustomActions: string[]
+): boolean => {
+  if (action.actionType === UserActionType.SEARCH && !action.query) {
+    return true;
+  }
+
+  if (action.actionType === UserActionType.CUSTOM) {
+    const eventType = action.eventData?.type || '';
+    const eventValue = action.eventData?.value || '';
+
+    return (
+      excludedCustomActions.includes(eventType) ||
+      excludedCustomActions.includes(eventValue)
+    );
+  }
+  return false;
+};
+
+/**
+ * Divides actions into sessions and organizes them in the timeline based on the ticket creation date. This function does the following:
+ * 1) Iterates over the actions and groups them into sessions based on the session inactivity threshold.
+ * 2) Inserts the case creation action in the current session.
+ * 3) Filters the actions in the current session to exclude custom actions where the type or value is included in the excludedCustomActions and also empty searches.
+ * 4) Inserts the session in the timeline at the right location.
+ * 5) Returns the timeline with the current session, 2 preceding sessions and 2 following sessions.
+ * @param actions {UserAction[]} - The actions to split
+ * @param ticketCreationDate {string} - The ticket creation date
+ * @param actionsToExclude {string[]}
+ * @returns {UserActionTimeline} - The timeline of user actions (current session and 2 preceding and following sessions)
+ */
+export const splitActionsIntoTimelineSessions = (
+  actions: UserAction[],
+  ticketCreationDate: number,
+  actionsToExclude: string[]
+): UserActionTimeline => {
+  const returnTimeline: UserActionTimeline = {
+    precedingSessions: [],
+    session: undefined,
+    followingSessions: [],
+  };
+
+  let currentSession: UserSession = {
+    start: actions[0].timestamp,
+    end: actions[0].timestamp,
+    actions: [],
+  };
+
+  actions.forEach((action) => {
+    if (isActionWithinSessionThreshold(action, currentSession.end)) {
+      currentSession.actions.push(action);
+      currentSession.start = action.timestamp;
+      return;
+    }
+
+    currentSession.actions = currentSession.actions.filter(
+      (action) => !shouldExcludeAction(action, actionsToExclude)
+    );
+
+    insertSessionInTimeline(currentSession, ticketCreationDate, returnTimeline);
```

Comment:
This doesn't fix the issue of inserting empty sessions to the timeline: https://github.com/coveo/ui-kit/pull/4223#discussion_r1722228253

Please add unit tests to cover that.